File: CompanionApps/iOS/Sources/Handlers/TouchHandler.swift
/// Handles touch-related gRPC requests by delegating to XCUITestBridge.
final class TouchHandler: @unchecked Sendable {
    private let bridge: XCUITestBridge

    init(bridge: XCUITestBridge) {
        self.bridge = bridge
    }

    func tap(x: Double, y: Double) async {
        await bridge.tap(x: x, y: y)
    }

    func doubleTap(x: Double, y: Double) async {
        await bridge.doubleTap(x: x, y: y)
    }

    func longPress(x: Double, y: Double, durationMs: Int) async {
        await bridge.longPress(x: x, y: y, durationSeconds: Double(durationMs) / 1000.0)
    }

    func tapElement(
        id: String?,
        label: String?,
        containsText: String? = nil,
        bundleID: String? = nil,
        candidateBundleIDs: [String] = []
    ) async -> Bool {
        await bridge.tapElement(
            id: id,
            label: label,
            containsText: containsText,
            bundleID: bundleID,
            candidateBundleIDs: candidateBundleIDs
        )
    }
}

File: CompanionApps/Android/app/src/androidTest/java/com/amoo/companion/handlers/TouchHandler.kt
package com.amoo.companion.handlers

import com.amoo.companion.bridge.UIAutomatorBridge

class TouchHandler(private val bridge: UIAutomatorBridge) {

    fun tap(x: Int, y: Int): Boolean = bridge.tap(x, y)

    fun doubleTap(x: Int, y: Int): Boolean {
        bridge.tap(x, y)
        Thread.sleep(50)
        return bridge.tap(x, y)
    }

    fun longPress(x: Int, y: Int, durationMs: Int): Boolean {
        return bridge.longPress(x, y, durationMs)
    }

    fun tapElement(
        resourceId: String?,
        text: String?,
        containsText: String? = null,
        appId: String? = null
    ): Boolean {
        // findTapTarget rather than findElements: on Compose the node that matches a label or text
        // is often not the node that can be clicked. See its documentation.
        val target = bridge.findTapTarget(resourceId, text, containsText, appId) ?: return false
        return bridge.tap(
            target.frame.x + target.frame.width / 2,
            target.frame.y + target.frame.height / 2
        )
    }
}
